kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/DataEncryptionKey.java

```diff
@@ -0,0 +1,333 @@
+                throw new IllegalArgumentException();
+            }
+            this.cipher = cipherSpec.newCipher();
+            this.paramSupplier = cipherSpec.paramSupplier();
+            this.cipherSpec = cipherSpec;
+            this.key = key;
+            this.numEncryptions = numEncryptions;
+        }
+
+        /**
+         * @return The encrypted key
+         */
+        public E edek() {
+            return edek;
+        }
+
+        /**
+         * Perform an encryption operation using the DEK.
+         * @param plaintext The plaintext to be encrypted
+         * @param aad The AAD to be included in the encryption
+         * @param paramAllocator A function that will return a buffer into which the cipher parameters will be written.
+         * The function's argument is the number of bytes required for the cipher parameters.
+         * @param ciphertextAllocator A function that will return a buffer into which the ciphertext will be written.
+         * The function's first argument is the number of bytes required for the cipher parameters.
+         * The function's second argument is the number of bytes required for the ciphertext.
+         * @throws DekUsageException If this Encryptor has run out of operations.
+         */
+        public void encrypt(@NonNull ByteBuffer plaintext,
+                            @Nullable ByteBuffer aad,
+                            @NonNull IntFunction<ByteBuffer> paramAllocator,
+                            @NonNull BiFunction<Integer, Integer, ByteBuffer> ciphertextAllocator) {
+
+            if (--numEncryptions >= 0) {
+                try {
+                    AlgorithmParameterSpec params = paramSupplier.get();
+                    cipher.init(Cipher.ENCRYPT_MODE, key, params);
+
+                    int size = cipherSpec.size(params);
+                    var parametersBuffer = paramAllocator.apply(size);
+                    cipherSpec.writeParameters(parametersBuffer, params);
+                    parametersBuffer.flip();
+
+                    int outSize = cipher.getOutputSize(plaintext.remaining());
+                    var ciphertext = ciphertextAllocator.apply(size, outSize);
+                    if (aad != null) {
+                        cipher.updateAAD(aad);
+                        aad.rewind();
+                    }
+                    var p = plaintext.position();
+                    cipher.doFinal(plaintext, ciphertext);
+                    plaintext.position(p);
+                    ciphertext.flip();
+
+                    if (numEncryptions == 0) {
+                        close();
+                    }
+                    return;
+                }
+                catch (GeneralSecurityException e) {
+                    throw new EncryptionException(e);
+                }
+            }
+            throw new DekUsageException("The Encryptor has no more operations allowed");
+        }
+
+        @Override
+        public void close() {
```

Comment:
Something for the future, we could potentially return unspent encryptions to the DEK. For example we've scrolled over a big batch and only encrypted 1 record in 1000, we could return 999 operations to the remaining.

Reply:
Yeah, I wondered about this too. It's not clear to me whether that really makes a difference. E.g. with record-based KEK selection we have to iterate over all the records in the request anyway; maintaining a per-KEK count doesn't add significant cost to that, but would allow us to request the exact number of encryptions we need for the second pass. And returning unused operations does come at the cost of more spinning on the atomic update required. So I left it out for now.